File: Lab5/C#/FigureLib/FigureLib/Rhombus.cs

using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace FigureLib
{
    public class Rhombus : Figures
    {
        private int XA, XB, XC, XD;
        private int YA, YB, YC, YD;
        public Rhombus(int XA, int YA, int XB, int YB,int XC, int YC,int XD, int YD)
        {
            this.XA = XA;
            this.YA = YA;   
            this.XB = XB;
            this.XC = XC;
            this.YC = YC;
            this.XD = XD;
            this.YD = YD;
        }

        public float calcDistance(int X1,int Y1,int X2, int Y2)
        {
            float d = MathF.Sqrt(MathF.Pow(X2 - X1, 2) + MathF.Pow(Y2 - Y1, 2));

            return d;
        }

        public override float calcArea()
        {
            float diag1 = calcDistance(XA, YA, XC, YC);
            float diag2 = calcDistance(XB, YB, XD, YD);

            float area = (diag1 * diag2) / 2;

            return area;
        }

        public override float calcPerimeter()
        {
            float[] sides = new float[4];
            sides[0] = calcDistance(XA, YA, XB, YB);
            sides[1] = calcDistance(XB, YB, XC, YC);
            sides[2] = calcDistance(XC, YC, XD, YD);
            sides[3] = calcDistance(XD, YD, XA, YA);

            float perimeter = sides[0] + sides[1] + sides[2] + sides[3];

            return perimeter;
        }
    }
}

File: Lab5/C++/FigureLib/FigureLib/Rhombus.cpp
#include "Rhombus.h"
#include <cmath>


Rhombus::Rhombus(int xA, int yA, int xB, int yB, int xC, int yC, int xD, int yD) : Figures() {
	this->xA = xA;
	this->xB = xB;
	this->xC = xC;
	this->xD = xD;
	this->yA = yA;
	this->yB = yB;
	this->yC = yC;
	this->yD = yD;
}

float Rhombus::calcDiagDistanse(int X1, int Y1, int X2, int Y2) {
	float d = sqrt(pow(X2 - X1, 2) + pow(Y2 - Y1, 2));

	return d;
}

float Rhombus::CalcArea() {
	float diag1, diag2;
	float area;

	diag1 = calcDiagDistanse(xA, yA, xC, yC);
	diag2 = calcDiagDistanse(xB, yB, xD, yD);

	area = 0.5 * diag1 * diag2;

	return area;
}

float Rhombus::CalcPerimeter() {
	float sides[4];

	sides[0] = calcDiagDistanse(this->xA, this->yA, this->xB, this->yB); // Сторона AB
	sides[1] = calcDiagDistanse(this->xB, this->yB, this->xC, this->yC); // Сторона BC
	sides[2] = calcDiagDistanse(this->xC, this->yC, this->xD, this->yD); // Сторона CD
	sides[3] = calcDiagDistanse(this->xD, this->yD, this->xA, this->yA); // Сторона DA


	float perimeter = sides[0] + sides[1] + sides[2] + sides[3];

	return perimeter;
}
